contracts/interchain-token/src/contract.rs

```diff
@@ -1,32 +1,365 @@
+        env.storage()
+            .instance()
+            .extend_ttl(INSTANCE_TTL_THRESHOLD, INSTANCE_TTL_EXTEND_TO);
+    }
+
+    fn extend_balance_ttl(env: &Env, key: &DataKey) {
+        env.storage()
+            .persistent()
+            .extend_ttl(key, INSTANCE_TTL_THRESHOLD, INSTANCE_TTL_EXTEND_TO);
+    }
+
+    fn read_allowance(env: &Env, from: Address, spender: Address) -> AllowanceValue {
+        let key = DataKey::Allowance(AllowanceDataKey { from, spender });
+        env.storage()
+            .temporary()
+            .get::<_, AllowanceValue>(&key)
+            .map_or(
+                AllowanceValue {
+                    amount: 0,
+                    expiration_ledger: 0,
+                },
+                |allowance| {
+                    if allowance.expiration_ledger < env.ledger().sequence() {
+                        AllowanceValue {
+                            amount: 0,
+                            expiration_ledger: allowance.expiration_ledger,
+                        }
+                    } else {
+                        allowance
+                    }
+                },
+            )
+    }
+
+    fn write_allowance(
+        env: &Env,
+        from: Address,
+        spender: Address,
+        amount: i128,
+        expiration_ledger: u32,
+    ) {
+        let allowance = AllowanceValue {
+            amount,
+            expiration_ledger,
+        };
+
+        assert_with_error!(
+            !(amount > 0 && expiration_ledger < env.ledger().sequence()),
+            "expiration_ledger is less than ledger seq when amount > 0"
+        );
+
+        let key = DataKey::Allowance(AllowanceDataKey { from, spender });
+        env.storage().temporary().set(&key, &allowance);
+
+        if amount > 0 {
+            let live_for = expiration_ledger
+                .checked_sub(env.ledger().sequence())
+                .unwrap();
+
+            env.storage()
+                .temporary()
+                .extend_ttl(&key, live_for, live_for)
+        }
+    }
+
+    fn spend_allowance(env: &Env, from: Address, spender: Address, amount: i128) {
+        let allowance = Self::read_allowance(env, from.clone(), spender.clone());
+
+        assert_with_error!(allowance.amount >= amount, "insufficient allowance");
+
+        if amount > 0 {
+            Self::write_allowance(
+                env,
+                from,
+                spender,
+                allowance
+                    .amount
+                    .checked_sub(amount)
+                    .expect("insufficient allowance"),
+                allowance.expiration_ledger,
+            );
+        }
+    }
+
+    fn read_balance(env: &Env, addr: Address) -> i128 {
+        let key = DataKey::Balance(addr);
+        env.storage()
+            .persistent()
+            .get::<DataKey, i128>(&key)
+            .map_or(0, |balance| {
+                Self::extend_balance_ttl(env, &key);
+                balance
+            })
```

Comment:
```suggestion
            .and_then(|balance| {
                Self::extend_balance_ttl(env, &key);
                balance
            })
            .unwrap_or_default()
```
a bit more readable maybe?

Reply:
```
error: using `Option.and_then(|x| Some(y))`, which is more succinctly expressed as `map(|x| y)`
   --> contracts/interchain-token/src/contract.rs:340:9
    |
340 | /         env.storage()
341 | |             .persistent()
342 | |             .get::<_, i128>(&key)
343 | |             .and_then(|balance| {
344 | |                 Self::extend_balance_ttl(env, &key);
345 | |                 Some(balance)
346 | |             })
    | |______________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bind_instead_of_map
    = note: `-D clippy::bind-instead-of-map` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::bind_instead_of_map)]`
help: try
    |
343 ~             .map(|balance| {
344 |                 Self::extend_balance_ttl(env, &key);
345 ~                 balance
    |

error: could not compile `interchain-token` (lib) due to 1 previous error
```

Reply:
and there's another compile error when using `map`
```
error: using `map` over `inspect`
   --> contracts/interchain-token/src/contract.rs:335:14
    |
335 |             .map(|balance| {
    |              ^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_inspect
    = note: `-D clippy::manual-inspect` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::manual_inspect)]`
help: try
    |
335 ~             .inspect(|balance| {
336 |                 // Extend the TTL of the balance entry when the balance is successfully retrieved.
337 ~                 Self::extend_balance_ttl(env, &key);
    |

error: could not compile `interchain-token` (lib) due to 1 previous error

```